README.md
# Bilderkennung mit Haar/Cascade (OpenCV)

## Haarcascade Verkehrszeichen

Erkennung und Klassifizierung von Straßenschildern mit Haarcascades

## Optionen

-d, --directory - Relativer Pfad zu den Eingabebildern. Wenn nicht angegeben, dann wird Video gestartet.

## Quellen

Basierend auf einem Video wurde grundlegende Struktur angelegt, siehe https://www.youtube.com/watch?v=zwiKIzvGAhM

Die Detektoren wurden nach den Schritten aus einem Video trainiert, siehe https://www.youtube.com/watch?v=POSYDLcspIk

Alles andere wurde mehr oder weniger mit Hilfe der OpenCV Dokumentation implementiert.


CHANGELOG.md
# Changelog

## v0.4.0 - 2023-01-08

- FPS
- Configuration
- Add files via upload
- Forms detection

## v0.3.0 - 2022-12-17

- Loop
- Imageconverter

## v0.2.0 - 2022-12-11

- Detector list

## v0.1.0 - 2022-11-21

- Haarcascade Frontal Face Recognition


main.py
from colors import colors
from detectors import detectors
from turtle import circle
import cv2
import numpy as np
import time
import argparse
import os

#Parameter initialisieren
parser = argparse.ArgumentParser( prog = 'Haarcascade Verkehrszeichen', description = 'Erkennung und Klassifizierung von Straßenschildern mit Haarcascades', epilog = '')
parser.add_argument('-d', '--directory', default='', help='Relativer Pfad zu den Eingabebildern. Wenn nicht angegeben, dann wird Video gestartet.' )

def detection(values, image):
    for value in values:
        for detector in detectors.values():
            if detector["form"] == value:
                faces = detector["cascade"].detectMultiScale(image_gray)

                for x, y, width, height in faces:
                    if width > 32:
                        cv2.rectangle(image, (x, y), (x + width, y + height), color = colors[detector["color"]], thickness = 2)
                        cv2.putText(image, detector["text"], (x, y - 5), font, 0.5, colors[detector["color"]], 2)

def timeInMS():
    return time.time() * 1000

# Parameter einlesen
args = parser.parse_args()

# camera / Directory initialisieren
bWithDir = True     # Flag ob Camera oder Directory Mode
capture = None      # Objekt für Zugriff auf Kamera
images = []         # Array mit den Pfaden zu den einzelnen Bildern (Directory Mode)
currentImg = 0      # Index des aktuell angezeigten Bildes (Directory Mode)

if args.directory == '':
    bWithDir = False
    capture = cv2.VideoCapture(0)
else:
    for filename in os.listdir(args.directory):
        f = os.path.join(args.directory, filename)
        if os.path.isfile(f):
            images.append(f)

# font initialisieren
font = cv2.FONT_HERSHEY_SIMPLEX

# time für Framerate initialisieren
last_time = timeInMS()

#Einlesen XML
for code, detector in detectors.items():
    detector["cascade"] = cv2.CascadeClassifier(f"detectors/{code}.detector.xml")

while True:
    contour_set = set()

    # Bild einlesen bzw. von der Kamera abgreifen (try/except, weil openCV mit äöüß im Dateinamen nicht klar kommt)
    try:
        if not bWithDir:
            _, image = capture.read()
        else:
            image = cv2.imread( images[currentImg] )

            # Bild skalieren
            scaleFactor = 1 # Faktor um den das Bild skaliert wird
            height = image.shape[0]
            width = image.shape[1]

            if height >= width:
                scaleFactor = 800 / height
            else:
                scaleFactor = 800 / width

            dim = (int(width*scaleFactor), int(height*scaleFactor)) # Neue Dimension des Bildes als Tuple
            resized = cv2.resize(image, dim, interpolation = cv2.INTER_AREA)

            image = resized
    except:
        break

    # Run only at 30 frames per second
    if timeInMS() > last_time + 33:
        image_gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

        # Lines
        ret, thresh = cv2.threshold(image_gray, 150, 255, cv2.THRESH_BINARY)
        contours, hierarchy = cv2.findContours(thresh, cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE)
        # Circle
        circles = cv2.HoughCircles(image_gray, cv2.HOUGH_GRADIENT,1.5,100, param1=300,param2=80,minRadius=1,maxRadius=0)

        # Detection Form
        for contour in contours:
               x,y,w,h = cv2.boundingRect(contour)
               if w > 32:
                approx = cv2.approxPolyDP(contour, 0.01*cv2.arcLength(contour,True), True)

                # Triangle
                if len(approx) == 3:
                    contour_set.add("triangle")

                # Square & rectangle
                elif len(approx) == 4:
                    (x,y, w, h) = cv2.boundingRect(approx)
                    ar = w / float(h)
                    if ar >= 0.95 and ar <= 1.05:
                        contour_set.add("square")
                    else:
                        contour_set.add("rectangle")

                # Octagon
                elif len(approx) == 8:
                    contour_set.add("octagon")

        # Circle
        if circles is not None:
            circles = np.round(circles[0, :]).astype("int")
            for (x, y, r) in circles:
                contour_set.add("circle")

    delta = timeInMS() - last_time
    last_time = timeInMS()
    fps = str(int(1000/delta))

    cv2.putText(image, fps, (10, 20), font, 0.5, (255, 255, 255), 1)

    detection(contour_set, image)

    cv2.imshow("Camera", image)

    # Tasteneingabe abgreifen
    key = cv2.waitKey(1)
    if key == ord("q"):
        break
    elif key == ord("n"):
        currentImg += 1
        if currentImg >= len(images):
            currentImg = 0

# Ressourcen freigeben
if not bWithDir:
    capture.release()
cv2.destroyAllWindows()


detectors.py
detectors = {
    "vz123": {
        "code": "vz123",
        "color": "green",
        "form": "triangle",
        "name": "Arbeitsstelle",
        "text": "Arbeitsstelle"
    },
    # "vz205": {
    #     "code": "vz205",
    #     "color": "red",
    #     "form": "triangle",
    #     "name": "Vorfahrt gewähren!",
    #     "text": "Vorfahrt gewaehren"
    # },
    "vz206": {
        "code": "vz206",
        "color": "red",
        "form": "octagon",
        "name": "Halt! Vorfahrt gewähren!",
        "text": "Stopp"
    },
    # "vz220-20": {
    #     "code": "vz220-20",
    #     "color": "teal",
    #     "form": "rectangle",
    #     "name": "Einbahnstraße",
    #     "text": "Einbahnstrasse"
    # },
    # "vz242": {
    #     "code": "vz242",
    #     "color": "pink",
    #     "form": "square",
    #     "name": "Fussgängerzone",
    #     "text": "Fussgaengerzone"
    # },
    "vz267": {
        "code": "vz267",
        "color": "purple",
        "form": "circle",
        "name": "Verbot der Einfahrt",
        "text": "Verbot der Einfahrt"
    },
    # "vz306": {
    #     "code": "vz306",
    #     "color": "blue",
    #     "form": "square",
    #     "name": "Vorfahrtstraße",
    #     "text": "Vorfahrtstrasse"
    # },
    "vz325": {
        "code": "vz325",
        "color": "orange",
        "form": "rectangle",
        "name": "Spielstraße",
        "text": "Spielstrasse"
    },
    "vz350": {
        "code": "vz350",
        "color": "orange",
        "form": "square",
        "name": "Fußgängerüberweg",
        "text": "Fusgaengerueberweg"
    }
}


uniprojekt.py
import cv2

capture = cv2.VideoCapture(0)
cascade = cv2.CascadeClassifier(r"vorfahrtdetector.xml")
cascade2 = cv2.CascadeClassifier(r"arbeitsstelledetector.xml")
cascade3 = cv2.CascadeClassifier(r"verbotdereinfahrtdetector.xml")

while True:
    _, im = capture.read()
    im_gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)
    font=cv2.FONT_HERSHEY_SIMPLEX
    faces = cascade.detectMultiScale(im_gray)
    faces2 = cascade2.detectMultiScale(im_gray)
    faces3 = cascade3.detectMultiScale(im_gray)
    for x, y, width, height in faces:
        cv2.rectangle(im, (x,y), (x + width,y + height), color=(0, 140, 255), thickness=3)
        cv2.putText(im,'Vorfahrt',(x,y),font,0.9,(0,140,255),2)

    for x, y, width, height in faces2:
        cv2.rectangle(im, (x,y), (x + width,y + height), color=(0, 215, 255), thickness=3)
        cv2.putText(im,'Arbeitsstelle',(x,y),font,0.9,(0,215,255),2)

    for x, y, width, height in faces3:
        cv2.rectangle(im, (x,y), (x + width,y + height), color=(0, 0, 255), thickness=3)
        cv2.putText(im,'Verbot der Einfahrt',(x,y),font,0.9,(0,0,255),2)

    cv2.imshow("Kamera", im)
    if cv2.waitKey(1) == ord("q"):
        break

capture.release()
cv2.destroyAllWindows()

imgConverter/imageconverter.py
import cv2
import os
import numpy as np
import argparse
from os.path import exists

# Parser für die CMD Parameter
parser = argparse.ArgumentParser( prog = 'PKIImageConverter', description = 'Konvertiert Bilder in ein einheitliches Format zum Anlernen von Bilderkennungsalgorithmen', epilog = '')
parser.add_argument('-f', '--file', default='' )
parser.add_argument('-d', '--directory', default='' )
parser.add_argument('-W', '--width', default='800', type=int )            # Height und Width mit Großem H und W, weil kleines h für help 
parser.add_argument('-H', '--height', default='800', type=int)       
parser.add_argument('-fb', '--fillBlack', default='True' , choices=['True', 'False'])
parser.add_argument('-o', '--output', default='' )
parser.add_argument('-op', '--outputPrefix', default='p' )
parser.add_argument('-ft', '--fileType', default='bmp' )


# Json zum verwalten der Settings - Defaultwerte können mit Aufrufparametern überschrieben werden
settings = {
    'width': 800,
    'height': 800,
    'fillWithBlack': True,
    'outputDir': '',    # Relativer Pfad zum Output Directory
    'outputFormat': 'bmp',
    'outputPrefix': 'p',
    'allowedInputFormats': ['.jpg', '.jpeg', '.png', '.bmp', '.webp', '.gif']
}

# Counter für die Benamung der Ausgabe Bilder
COUNTER = 1


# Funktion zum konvertieren von Bildern
def convertImage(inputImg, counter):
    if os.path.isfile(inputImg):
        # Name und Pfad der Ausgabe
        outputImg = os.path.join(settings['outputDir'], settings['outputPrefix'] + str(counter).zfill(3) + '.' + settings['outputFormat'])
        # Counter hochzählen, wenn Bild mit diesem Namen schon vorhanden ist
        while exists(outputImg):
            counter += 1
            outputImg = os.path.join(settings['outputDir'], settings['outputPrefix'] + str(counter).zfill(3) + '.' + settings['outputFormat'])
   
        # Try Except, weil teilweise Bilder von OpenCV nicht gelesen werden 
        try:
            # Bild einlesen
            img = cv2.imread(inputImg)
            
            # Bild skalieren
            scaleFactor = 1 # Faktor um den das Bild skaliert wird
            height = img.shape[0]
            width = img.shape[1]
            
            if height >= width:
                scaleFactor = settings['height'] / height
            else:
                scaleFactor = settings['width'] / width
                
            dim = (int(width*scaleFactor), int(height*scaleFactor)) # Neue Dimension des Bildes als Tuple
            resized = cv2.resize(img, dim, interpolation = cv2.INTER_AREA)
                        
            # schwarz auffüllen
            if settings['fillWithBlack']:
                offsetWidth = int( (settings['width'] - dim[0]) / 2 )
                offsetHeight = int( (settings['height'] - dim[1]) / 2 )
                            
                if offsetHeight > 0 or offsetWidth > 0:
                    # Leeres Schwarzes Bild erstellen
                    blackedImg = np.zeros((settings['height'], settings['width'],3), np.uint8) 
                    blackedImg[:,:] = (0,0,0)
                    
                    # In das Schwarze bild wird Mittig über die Offsets das eigentliche Bild reinkopiert
                    tmpImg = blackedImg.copy()
                    tmpImg[ offsetHeight:offsetHeight+dim[1], offsetWidth:offsetWidth+dim[0] ] = resized.copy()
                    resized = tmpImg
                         
            # Bild schreiben
            cv2.imwrite(outputImg, resized)
        except:
            print('FEHLER: ' + inputImg + ' konnte nicht konvertiert werden!')
    
'''
for filename in os.listdir('input'):
    f = os.path.join('input', filename)
    if os.path.isfile(f):
        convertImage(f, COUNTER)
        COUNTER += 1
'''        

if __name__ == '__main__':
    args = parser.parse_args()
    settings['height'] = args.height
    settings['width'] = args.width
    settings['fillWithBlack'] = eval(args.fillBlack)
    settings['outputDir'] = args.output
    settings['outputFormat'] = args.fileType
    settings['outputPrefix'] = args.outputPrefix
    
    if args.file != '':
        convertImage(inputImg, 1)
        
    if args.directory != '':
        for filename in os.listdir(args.directory):
            if os.path.splitext(filename)[1].lower() in settings['allowedInputFormats']:
                f = os.path.join(args.directory, filename)
                if os.path.isfile(f):
                    convertImage(f, COUNTER)
                    COUNTER += 1





haarcascade.py
from turtle import circle
import cv2
import numpy as np
import time
import argparse
import os

#Parameter initialisieren
parser = argparse.ArgumentParser( prog = 'Haarcascade Verkehrszeichen', description = 'Erkennung und Klassifizierung von Straßenschildern mit Haarcascades', epilog = '')
parser.add_argument('-d', '--directory', default='', help='Relativer Pfad zu den Eingabebildern. Wenn nicht angegeben, dann wird Video gestartet.' )

def detection(values, image):
    for value in values:
        for detector in detectors.values():
            if detector["form"] == value:
                faces = detector["cascade"].detectMultiScale(image_gray)

                for x, y, width, height in faces:
                    if width > 50:
                        cv2.rectangle(image, (x, y), (x + width, y + height), color = detector["color"], thickness = 2)
                        cv2.putText(image, detector["text"], (x, y - 5), font, 0.5, detector["color"], 2)

def timeInMS():
    return time.time() * 1000

detectors = {
    "vz123": {
        "code": "vz123",
        "color": (50, 200, 50),
        "form": "triangle",
        "name": "Arbeitsstelle",
        "text": "Arbeitsstelle"
    },
    "vz205": {
        "code": "vz205",
        "color": (50, 50, 200),
        "form": "triangle",
        "name": "Vorfahrt gewähren!",
        "text": "Vorfahrt gewaehren"
    },
    "vz206": {
        "code": "vz206",
        "color": (50, 50, 200),
        "form": "octagon",
        "name": "Halt! Vorfahrt gewähren!",
        "text": "Stopp"
    },
    "vz267": {
        "code": "vz267",
        "color": (200, 50, 150),
        "form": "circle",
        "name": "Verbot der Einfahrt",
        "text": "Verbot der Einfahrt"
    },
    "vz306": {
        "code": "vz306",
        "color": (200, 50, 50),
        "form": "square",
        "name": "Vorfahrtstraße",
        "text": "Vorfahrtstrasse"
    },
    "vz350": {
        "code": "vz350",
        "color": (50, 150, 250),
        "form": "square",
        "name": "Fußgängerüberweg",
        "text": "Fusgaengerueberweg"
    },
    "vz220-20": {
        "code": "vz220-20",
        "color": (115, 160, 230),
        "form": "rectangle",
        "name": "Einbahnstrasse",
        "text": "Einbahnstrasse"
    },
        "vz325": {
        "code": "vz325",
        "color": (120, 230, 115),
        "form": "rectangle",
        "name": "Spielstrasse",
        "text": "Spielstrasse"
    },
        "vz242": {
        "code": "vz242",
        "color": (230, 100, 120),
        "form": "square",
        "name": "Fussgaengerzone",
        "text": "Fussgaengerzone"
    } 
}

# Parameter einlesen
args = parser.parse_args()

# camera / Directory initialisieren
bWithDir = True     # Flag ob Camera oder Directory Mode
capture = None      # Objekt für Zugriff auf Kamera
images = []         # Array mit den Pfaden zu den einzelnen Bildern (Directory Mode)
currentImg = 0      # Index des aktuell angezeigten Bildes (Directory Mode)

if args.directory == '':
    bWithDir = False
    capture = cv2.VideoCapture(0)
else:
    for filename in os.listdir(args.directory):
        f = os.path.join(args.directory, filename)
        if os.path.isfile(f):
            images.append(f) 

# font initialisieren
font = cv2.FONT_HERSHEY_SIMPLEX

# time für Framerate initialisieren
last_time = timeInMS()

#Einlesen XML
for code, detector in detectors.items():
    detector["cascade"] = cv2.CascadeClassifier(f"detectors/{code}.detector.xml")

while True:
    # Bild einlesen bzw. von der Kamera abgreifen (try/except, weil openCV mit äöüß im Dateinamen nicht klar kommt)
    try:
        if not bWithDir:
            _, image = capture.read()
        else:
            image = cv2.imread( images[currentImg] )
    except:
        break

    # Run only at 30 frames per second
    if timeInMS() > last_time + 1:
        image_gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

        # Lines
        ret, thresh = cv2.threshold(image_gray, 150, 255, cv2.THRESH_BINARY)
        contours, hierarchy = cv2.findContours(thresh, cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE)
        # Circle
        circles = cv2.HoughCircles(image_gray, cv2.HOUGH_GRADIENT,1.5,100, param1=300,param2=80,minRadius=1,maxRadius=0)

        contour_set = set()
        # Detection Form
        for contour in contours:
               x,y,w,h = cv2.boundingRect(contour)
               if w > 10:
                approx = cv2.approxPolyDP(contour, 0.01*cv2.arcLength(contour,True), True)

                # Triangle
                if len(approx) == 3:
                    contour_set.add("triangle")

                # Square & rectangle
                elif len(approx) == 4:
                    (x,y, w, h) = cv2.boundingRect(approx)
                    ar = w / float(h)
                    if ar >= 0.95 and ar <= 1.05:
                        contour_set.add("square")
                    else:
                        contour_set.add("rectangle")

                # Octagon
                elif len(approx) == 8:
                    contour_set.add("octagon")

        # Circle
        if circles is not None:
            circles = np.round(circles[0, :]).astype("int")
            for (x, y, r) in circles:
                contour_set.add("circle")

    detection(contour_set, image)
    last_time = timeInMS()

    cv2.imshow("Camera", image)

    # Tasteneingabe abgreifen
    key = cv2.waitKey(1)
    if key == ord("q"):
        break
    elif key == ord("n"):
        currentImg += 1
        if currentImg >= len(images):
            currentImg = 0

# Ressourcen freigeben
if not bWithDir:
    capture.release()
cv2.destroyAllWindows()